### PR TITLE
Add support to names keyword in Index

### LINKED ISSUE #19082: BUG: Creating Index name using `names` names argument, doesn't set index name
#### Code Sample, a copy-pastable example if possible

```python
# This is true
i = pd.Index([], name='a')
assert i.name == 'a'
assert i.names == ('a',)

# This is false (fails assertion)
i = pd.Index([], names=('a',))
assert i.name == 'a'
assert i.names == ('a',)
```
#### Problem description

Setting Index name using `name` parameter works correctly, but using `names` parameter doesn;t do anything, i.e. `i.name is None`. While the documentation doesn't list `names` as a parameter, since it is an attribute, it gets listed as a parameter, e.g. in IDE.  

#### Expected Output
```python
# This is true
i = pd.Index([], name='a')
assert i.name == 'a'
assert i.names == ('a',)

# This is also true
i = pd.Index([], names=('a',))
assert i.name == 'a'
assert i.names == ('a',)
```

#### Output of ``pd.show_versions()``

<details>
INSTALLED VERSIONS
------------------
commit: None
python: 3.6.3.final.0
python-bits: 64
OS: Windows
OS-release: 10
machine: AMD64
processor: Intel64 Family 6 Model 94 Stepping 3, GenuineIntel
byteorder: little
LC_ALL: None
LANG: None
LOCALE: None.None

pandas: 0.22.0
pytest: 3.3.1
pip: 9.0.1
setuptools: 36.5.0.post20170921
Cython: 0.26.1
numpy: 1.13.3
scipy: 0.19.1
pyarrow: 0.7.1
xarray: None
IPython: 6.1.0
sphinx: 1.6.3
patsy: 0.4.1
dateutil: 2.6.1
pytz: 2017.2
blosc: None
bottleneck: 1.2.1
tables: 3.4.2
numexpr: 2.6.2
feather: 0.4.0
matplotlib: 2.1.0
openpyxl: 2.4.8
xlrd: 1.1.0
xlwt: 1.3.0
xlsxwriter: 1.0.2
lxml: 4.1.0
bs4: 4.6.0
html5lib: 0.999999999
sqlalchemy: 1.1.13
pymysql: None
psycopg2: None
jinja2: 2.9.6
s3fs: None
fastparquet: 0.1.3
pandas_gbq: 0.2.1
pandas_datareader: 0.5.0
</details>

### PR DESCRIPTION
**Pull request by pedrocunial**
- [x] closes #19082
- [x] tests added / passed
- [x] passes `black pandas`
- [x] passes `git diff upstream/master -u -- "*.py" | flake8 --diff`
- [ ] whatsnew entry

### CODE REVIEW

File: pandas/core/indexes/base.py [hunk 1]
```diff
@@ -259,12 +261,22 @@ def __new__(
         dtype=None,
         copy=False,
         name=None,
+        names=None,
```
Comment (WillAyd) [inline pandas/core/indexes/base.py]: Can you move this to the end (before kwargs)

File: pandas/core/indexes/base.py [hunk 2]
```diff
@@ -259,12 +261,22 @@ def __new__(
...
         fastpath=None,
         tupleize_cols=True,
         **kwargs
     ):
 
-        if name is None and hasattr(data, "name"):
+        if names is not None:
```
Comment (WillAyd) [inline pandas/core/indexes/base.py]: Not super familiar with this part of the code but is it possible to share the length of checking of `names` across the index types? This already fails for a MultiIndex:

```python
>>> pd.MultiIndex.from_tuples((('a', 'b'), ('c', 'd')), names=('a', 'b', 'c'))

ValueError: Length of names must match number of levels in MultiIndex.
```

So ideally could share that machinery here

**Review by WillAyd — CHANGES_REQUESTED** [review]

**Comment by TomAugspurger** [comment]
> With that said I'm a little hesitant on this change. I understand the goal but it feels weird to have both a name and names argument as part of the constructor.

I'd say that's an issue with the `Index` constructors API. They're both valid arguments, this just updating the signature / docs to reflect that :)

**Comment by WillAyd** [comment]
> They're both valid arguments, this just updating the signature / docs to reflect that :)

I assume that's just because the Index constructor appears to accept anything though, right? Not necessarily that it's valid.

```python
>>> pd.Index([], names=["foo"])
Index([], dtype='object')
>>> pd.Index([], names=["foo"])
Index([], dtype='object')
>>> pd.Index([], something="foo")
Index([], dtype='object')
```

**Comment by TomAugspurger** [comment]
It's used if you take the MultiIndex route

```python
In [5]: pd.Index([(1, 2), (3, 4)], names=['a', 'b'], tupleize_cols=True)
Out[5]:
MultiIndex([(1, 2),
            (3, 4)],
           names=['a', 'b'])
```

**Comment by WillAyd** [comment]
Sorry I think I went in a circle but coming back to it...I don't see how having both `name` and `names` makes sense from an API perspective. I get we want to do that to achieve symmetry in our inheritance but from an end user perspective this feels like a step in the wrong direction

**Comment by jreback** [comment]
closing; we want to deprecate names I think, this is going backwards.

**Pull request closed without merging**